misk-crypto/src/main/kotlin/misk/crypto/FieldLevelEncryptionPacket.kt

```diff
@@ -0,0 +1,358 @@
+      }
+
+      override fun type() = type
+    }
+
+    data class PayloadSized(override val payload: ByteArray) : Field(payload) {
+      companion object {
+        val type: EntryType = EntryType.PAYLOAD_SIZED
+      }
+
+      override fun type() = type
+    }
+
+    data class PayloadRemainder(override val payload: ByteArray) : Field(payload) {
+      companion object {
+        val type: EntryType = EntryType.PAYLOAD_REMAINDER
+      }
+
+      override fun type() = type
+
+      override fun writeToStreamIfNotEmpty(s: DataOutput) {
+        if (payload.isEmpty())
+          return
+
+        s.writeByte(this.type().type)
+        s.write(payload)
+      }
+    }
+
+    abstract fun type(): EntryType
+
+    open fun writeToStreamIfNotEmpty(s: DataOutput) {
+      if (payload.isEmpty())
+        return
+
+      if (payload.size > Short.MAX_VALUE) {
+        throw InvalidEncryptionContextException("Field is too large (${payload.size} is greater than ${Short.MAX_VALUE})")
+      }
+
+      s.writeByte(this.type().type)
+      s.writeShort(payload.size)
+      s.write(payload)
+    }
+
+    companion object {
+      fun readField(s: DataInputStream): Field? {
+        val type = s.read()
+        if (type < 0)
+          return null
+        return when (type) {
+          ExpandedContextDescription.type.type -> {
+            val size = s.readUnsignedShort()
+            val payload = ByteArray(size)
+            s.readFully(payload, 0, size)
+            ExpandedContextDescription(payload)
+          }
+
+          CompleteAad.type.type -> {
+            val size = s.readUnsignedShort()
+            val payload = ByteArray(size)
+            s.readFully(payload, 0, size)
+            CompleteAad(payload)
+          }
+
+          PayloadSized.type.type -> {
+            val size = s.readUnsignedShort()
+            val payload = ByteArray(size)
+            s.readFully(payload, 0, size)
+            PayloadSized(payload)
+          }
+
+          PayloadRemainder.type.type -> {
+            val stream = ByteArrayOutputStream()
+            var byteRead: Int
+            while (true) {
+              byteRead = s.read()
+              if (byteRead < 0)
+                break
+              stream.write(byteRead)
+            }
+            PayloadRemainder(stream.toByteArray())
+          }
+          else -> throw InvalidEncryptionContextException("Bad packet format, unknown field type: $type")
+        }
+      }
+    }
+  }
+
+  class Builder {
+    private var context = CaseInsensitiveMap<String, String>()
+
+    /**
+     * Add an entry to the context that can have an explicit value associated with it.
+     */
+    fun addContextEntry(key: String, value: String) = also {
+      if (key in context) throw InvalidEncryptionContextException("trying to add a context value that already exists ($key)")
+      if (key.toUpperCase() == ContextKey.UNDEFINED.name) throw InvalidEncryptionContextException("undefined is a reserved entry")
```

Comment:
toUpperCase() doesn’t do what you want. You want `toUpperCase(Locale.US)`
http://mattryall.net/blog/2009/02/the-infamous-turkish-locale-bug